src/Canvas.cc

```diff
@@ -782,57 +773,57 @@ Canvas::GetWeightFromCSSString(const char *weight) {
-
   // One of the user-specified families could map to multiple SFNT family names
   // if someone registered two different fonts under the same family name.
   // https://drafts.csswg.org/css-fonts-3/#font-style-matching
-  char **families = g_strsplit(pango_font_description_get_family(desc), ",", -1);
-  GHashTable *seen_families = g_hash_table_new(g_str_hash, g_str_equal);
-  GString *resolved_families = g_string_new("");
-
-  for (int i = 0; families[i]; ++i) {
-    GString *renamed_families = g_string_new("");
-    std::vector<FontFace>::iterator it = _font_face_list.begin();
-
-    for (; it != _font_face_list.end(); ++it) {
-      if (g_ascii_strcasecmp(families[i], pango_font_description_get_family(it->user_desc)) == 0) {
-        char *name = g_strdup(pango_font_description_get_family(it->sys_desc));
-        bool unseen = g_hash_table_lookup(seen_families, name) == NULL;
+  FontFace best;
+  istringstream families(pango_font_description_get_family(desc));
+  unordered_set<string> seen_families;
+  string resolved_families;
+  bool first = true;
+
+  for (string family; getline(families, family, ','); ) {
+    string renamed_families;
+    for (auto& ff : _font_face_list) {
+      string pangofamily = string(pango_font_description_get_family(ff.user_desc));
+      if (streq_casein(family, pangofamily)) {
+        const char* sys_desc_family_name = pango_font_description_get_family(ff.sys_desc);
+        bool unseen = seen_families.find(sys_desc_family_name) == seen_families.end();
 
         // Avoid sending duplicate SFNT font names due to a bug in Pango for macOS:
         // https://bugzilla.gnome.org/show_bug.cgi?id=762873
         if (unseen) {
-          g_hash_table_replace(seen_families, name, name);
-          if (renamed_families->len) g_string_append(renamed_families, ",");
-          g_string_append(renamed_families, name);
+          seen_families.insert(sys_desc_family_name);
+          if (renamed_families.size()) renamed_families += ',';
+          renamed_families += sys_desc_family_name;
         }
 
-        if (i == 0 && (best.user_desc == NULL || pango_font_description_better_match(desc, best.user_desc, it->user_desc))) {
-          best = *it;
+        if (first && (best.user_desc == nullptr || pango_font_description_better_match(desc, best.user_desc, ff.user_desc))) {
+          best = ff;
         }
       }
     }
 
-    if (resolved_families->len) g_string_append(resolved_families, ",");
-    g_string_append(resolved_families, renamed_families->len ? renamed_families->str : families[i]);
-    g_string_free(renamed_families, true);
+    if (resolved_families.size()) resolved_families += ',';
+    resolved_families += renamed_families.size() ? renamed_families : family;
+    first = false;
   }
 
-  ret = pango_font_description_copy(best.sys_desc ? best.sys_desc : desc);
-  pango_font_description_set_family_static(ret, resolved_families->str);
-
-  g_strfreev(families);
-  g_string_free(resolved_families, false);
```

Comment:
I forget exactly, but this might have been one leak (perhaps only when setting the font to a value that has already been used before). I don't think `resolved_families`'s character data is ever free'd.

Reply:
Yep. The `false` is because due to `pango_font_description_set_family_static`, I wanted to share the same pointer, but forgot to free it somewhere.